File: src/emanesh/emanesh/events/tdmascheduleevent.py
#
# Copyright (c) 2015 - Adjacent Link LLC, Bridgewater, New Jersey
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
# * Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in
#   the documentation and/or other materials provided with the
#   distribution.
# * Neither the name of Adjacent Link LLC nor the names of its
#   contributors may be used to endorse or promote products derived
#   from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#

from . import Event

import tdmascheduleevent_pb2
from collections import namedtuple

class TDMAScheduleEvent(Event):
    IDENTIFIER = 105
    TX=1
    RX=2
    IDLE=3

    _FrameEntry = namedtuple('FrameEntry', ['frame', 'slots'])

    def __init__(self,**kwargs):
        self._event = tdmascheduleevent_pb2.TDMAScheduleEvent()
        self._frames = {}

        for (name,value) in kwargs.items():
            if name == 'frequency':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                    value >= 0:
                    self._event.frequencyHz = value
                else:
                    raise ValueError("frequency must be a positive numeric")

            elif name == 'datarate':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                    value >= 0:
                    self._event.dataRatebps = value
                else:
                    raise ValueError("datarate must be a positive numeric")

            elif name == 'service':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                    value >= 0:
                    self._event.serviceClass = value
                else:
                    raise ValueError("service must be a positive numeric")

            elif name == 'power':
                if (isinstance(value,int) or \
                    isinstance(value,long) or \
                    isinstance(value,float)):
                    self._event.powerdBm = value
                else:
                    raise ValueError("power must be a numeric")

            else:
                raise KeyError("unknown parameter: %s" % name)

    def structure(self,**kwargs):
        slotsPerFrame = None
        framesPerMultiFrame = None
        slotOverheadMicroseconds = None
        slotDurationMicroseconds = None
        bandwidthHz = None

        if not kwargs:

            if self._event.HasField('structure'):
                return {'slots': self._event.structure.slotsPerFrame,
                        'frames': self._event.structure.framesPerMultiFrame,
                        'slotduration':self._event.structure.slotDurationMicroseconds,
                        'slotoverhead':self._event.structure.slotOverheadMicroseconds,
                        'bandwidth': self._event.structure.bandwidthHz}
            else:
                return None

        for (name,value) in kwargs.items():
            if name == 'slots':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                    value > 0:
                    slotsPerFrame = value
                else:
                    raise ValueError("'slots' must be a positive integer greater than 0")

            elif name == 'frames':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                    value > 0:
                    framesPerMultiFrame = value
                else:
                    raise ValueError("'frames' must be a positive integer greater than 0")

            elif name == 'slotduration':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                    value > 0:
                    slotDurationMicroseconds = value
                else:
                    raise ValueError("'slotduration' must be a positive integer greater than 0")

            elif name == 'slotoverhead':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                    value >= 0:
                    slotOverheadMicroseconds = value
                else:
                    raise ValueError("'slotoverhead' must be a positive integer (usecs)")

            elif name == 'bandwidth':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                    value > 0:
                    bandwidthHz = value
                else:
                    raise ValueError("'bandwidth' must be a positive integer greater than 0 (Hz)")

            else:
                raise KeyError("unknown parameter: %s" % name)


        if slotsPerFrame == None or \
           framesPerMultiFrame == None or \
           slotOverheadMicroseconds == None or \
           slotDurationMicroseconds == None or \
           bandwidthHz == None:
            raise KeyError("Missing one ore more keys: 'slots', 'frames', 'slotduration', 'slotoverhead', 'bandwidth'")

        self._event.structure.slotsPerFrame = slotsPerFrame
        self._event.structure.framesPerMultiFrame = framesPerMultiFrame
        self._event.structure.slotDurationMicroseconds = slotDurationMicroseconds
        self._event.structure.slotOverheadMicroseconds = slotOverheadMicroseconds
        self._event.structure.bandwidthHz = bandwidthHz


    def append(self,frameIndex,slotIndex,**kwargs):
        frameFrequencyHz = None
        frameDataRatebps = None
        frameClass = None
        framePowerdBm = None
        slotFrequencyHz = None
        slotDataRatebps = None
        slotClass = None
        slotPowerdBm = None
        slotType = None
        slotDestination = None

        if frameIndex in self._frames and \
           slotIndex in self._frames[frameIndex].slots:
            raise ValueError("slot index already defined for frame")

        for (name,value) in kwargs.items():
            if name == 'frame.frequency':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                value > 0:
                    frameFrequencyHz = value
                else:
                    raise ValueError("'frame.frequency' must be a integer greater that 0 (Hz)")

            elif name == 'frame.datarate':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                value > 0:
                    frameDataRatebps = value
                else:
                    raise ValueError("'frame.datarate' must be a positive integer greater than 0 (bps)")

            elif name == 'frame.service':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                value >= 0 and value <= 3:
                    frameClass = value
                else:
                    raise ValueError("'frame.service' must be a positive integer in the set [0,3]")

            elif name == 'frame.power':
                if (isinstance(value,int) or \
                    isinstance(value,long) or \
                    isinstance(value,float)):

                    framePowerdBm = value
                else:
                    raise ValueError("'frame.power' must be a numeric (dBm)")

            elif name == 'frequency':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                value > 0:
                    slotFrequencyHz = value
                else:
                    raise ValueError("'frequency' must be a integer greater that 0 (Hz)")

            elif name == 'datarate':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                value > 0:
                    slotDataRatebps = value
                else:
                    raise ValueError("'datarate' must be a positive integer greater than 0 (bps)")

            elif name == 'service':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                value >= 0 and value <= 3:
                    slotClass = value
                else:
                    raise ValueError("'service' must be a positive integer in the set [0,3]")

            elif name == 'power':
                if (isinstance(value,int) or \
                    isinstance(value,long) or \
                    isinstance(value,float)):
                    slotPowerdBm = value
                else:
                    raise ValueError("'power' must be a numeric (dBm)")


            elif name == 'destination':
                if (isinstance(value,int) or \
                    isinstance(value,long)) and \
                value > 0:
                    slotDestination = value
                else:
                    raise ValueError("'destination' must be a positive integer (NEM Id)")

            elif name == 'type':
                if value == "tx" or value == TDMAScheduleEvent.TX:
                    slotType = "tx"

                elif value == "rx" or value == TDMAScheduleEvent.RX:
                    slotType = "rx"

                elif value =="idle" or value == TDMAScheduleEvent.IDLE:
                    slotType = "idle"

                else:
                    raise ValueError("'type' must be one of: tx, rx or idle")

            else:
                raise KeyError("unknown parameter: %s" % name)

        if slotType == "tx":
            if slotFrequencyHz == None and \
               frameFrequencyHz == None and \
               self._event.frequencyHz == None:
                raise KeyError("tx slot 'frequency' must be specified when 'frame.frequency' missing and default not set")

            if slotDataRatebps == None and \
               frameDataRatebps == None and \
               self._event.dataRatebps == None:
                raise KeyError("tx slot 'datarate' must be specified when 'frame.datarate' missing and default not set")

            if slotClass == None and \
               frameClass == None and \
               self._event.serviceClass == None:
                raise KeyError("tx slot 'service' must be specified when 'frame.service' missing and default not set")

            if slotPowerdBm != None and \
               framePowerdBm == None and \
               self._event.powerdBm == None:
                raise KeyError("tx slot 'power' must be specified when 'frame.power' missing and default not set")

        elif slotType == "rx":
            if slotDataRatebps != None or \
               slotClass != None or \
               slotPowerdBm != None or \
               slotDestination != None:
                raise KeyError("rx slot cannot have 'datarate', 'service', 'power' and/or 'destination'")

            if slotFrequencyHz == None and \
               frameFrequencyHz == None and \
               self._event.frequencyHz == None:
                raise KeyError("rx slot 'frequency' must be specified when 'frame.frequency' missing and default not set")

        elif slotType == "idle":
            if slotFrequencyHz != None or \
               slotDataRatebps != None or \
               slotClass != None or \
               slotPowerdBm != None:
                raise ValueError("idle slot cannot have 'frequency', 'datarate', 'service', 'power',  and/or 'destination'")

        else:
            raise KeyError("missing 'type'")

        if frameIndex in self._frames:
            frame = self._frames[frameIndex].frame
        else:
            frame = self._event.frames.add()
            self._frames[frameIndex] = TDMAScheduleEvent._FrameEntry(frame,set())
            frame.index = frameIndex

        if frameFrequencyHz != None:
            frame.frequencyHz = frameFrequencyHz

        if frameDataRatebps != None:
            frame.dataRatebps = frameDataRatebps

        if frameClass != None:
            frame.serviceClass = frameClass

        if framePowerdBm != None:
            frame.powerdBm = framePowerdBm

        slot = frame.slots.add()

        slot.index = slotIndex

        if slotType == "tx":
            slot.type = tdmascheduleevent_pb2.TDMAScheduleEvent.Frame.Slot.SLOT_TX

            if slotFrequencyHz != None:
                slot.tx.frequencyHz = slotFrequencyHz

            if slotDataRatebps != None:
                slot.tx.dataRatebps = slotDataRatebps

            if slotClass != None:
                slot.tx.serviceClass = slotClass

            if slotPowerdBm != None:
                slot.tx.powerdBm = slotPowerdBm

            if slotDestination != None:
                slot.tx.destination = slotDestination

        elif slotType == "rx":
            slot.type = tdmascheduleevent_pb2.TDMAScheduleEvent.Frame.Slot.SLOT_RX

            if slotFrequencyHz != None:
                slot.rx.frequencyHz = slotFrequencyHz

        else:
            slot.type = tdmascheduleevent_pb2.TDMAScheduleEvent.Frame.Slot.SLOT_IDLE

        self._frames[frameIndex].slots.add(slotIndex)

    def serialize(self):
        return self._event.SerializeToString()

    def restore(self,data):
        self._event.ParseFromString(data)

    def __iter__(self):
        for frame in self._event.frames:
            kwargs = {'index' : frame.index}

            if frame.HasField('frequencyHz'):
                kwargs['frame.frequency'] = frame.frequencyHz

            if frame.HasField('dataRatebps'):
                kwargs['frame.datarate'] = frame.dataRatebps

            if frame.HasField('serviceClass'):
                kwargs['frame.service'] = frame.serviceClass

            if frame.HasField('powerdBm'):
                kwargs['frame.power'] = frame.powerdBm

            slots = {}

            for slot in frame.slots:
                s = {}

                if slot.type == tdmascheduleevent_pb2.TDMAScheduleEvent.Frame.Slot.SLOT_TX:
                    s['type'] = 'tx'

                    if slot.HasField('tx'):
                        if slot.tx.HasField('frequencyHz'):
                            s['frequency'] = slot.tx.frequencyHz

                        if slot.tx.HasField('dataRatebps'):
                            s['datarate'] = slot.tx.dataRatebps

                        if slot.tx.HasField('serviceClass'):
                            s['service'] = slot.tx.serviceClass

                        if slot.tx.HasField('powerdBm'):
                            s['power'] = slot.tx.powerdBm

                        if slot.tx.HasField('destination'):
                            s['destination'] = slot.tx.destination

                elif slot.type == tdmascheduleevent_pb2.TDMAScheduleEvent.Frame.Slot.SLOT_RX:
                    s['type'] = 'rx'

                    if slot.HasField('rx') and slot.rx.HasField('frequencyHz'):
                        s['frequency'] = slot.rx.frequencyHz

                else:
                    s['type'] = 'idle'

                slots[slot.index] = s

            kwargs['slots'] = slots

            yield kwargs


if __name__ == "__main__":
    import sys
    from optparse import OptionParser
    from emanesh.events import EventService
    import copy
    from . import TDMASchedule

    usage="%prog [OPTION]... NEMID..."
    description=""
    epilog=""

    class LocalParser(OptionParser):
        def format_epilog(self, formatter):
            return self.epilog

    optionParser = LocalParser(usage=usage,
                               description=description,
                               epilog=epilog)

    optionParser.add_option("-p",
                            "--port",
                            action="store",
                            type="int",
                            dest="port",
                            default=45703,
                            help="Event channel listen port [default: %default]")

    optionParser.add_option("-g",
                            "--group",
                            action="store",
                            type="string",
                            dest="group",
                            default="224.1.2.8",
                            help="Event channel multicast group [default: %default]")

    optionParser.add_option("-i",
                            "--device",
                            action="store",
                            type="string",
                            dest="device",
                            help="Event channel multicast device")

    optionParser.add_option("-f",
                            "--file",
                            action="store",
                            type="string",
                            dest="file",
                            help="Schedule XML file")

    (options, args) = optionParser.parse_args()

    if options.file:
        schedule = TDMASchedule(options.file)
    else:
        print >>sys.stderr, "currently only --file supported"
        exit(1)

    eventService = EventService((options.group,options.port,options.device))

    info = schedule.info()

    structure = schedule.structure()

    frameDefaults =  schedule.defaultsFrame();

    for nodeId in info:
        event = TDMAScheduleEvent(**schedule.defaults())
        for frameIndex in info[nodeId]:
            for slotIndex,args in info[nodeId][frameIndex].items():
                defaults = args
                for key,value in frameDefaults[frameIndex].items():
                    defaults["frame." + key] = value
                event.append(frameIndex,slotIndex,**defaults)

        if structure != None:
            event.structure(**structure)

        eventService.publish(nodeId,event)

File: src/libemane/tdmascheduleevent.cc
/*
 * Copyright (c) 2015-2016 - Adjacent Link LLC, Bridgewater, New Jersey
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * * Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in
 *   the documentation and/or other materials provided with the
 *   distribution.
 * * Neither the name of Adjacent Link LLC nor the names of its
 *   contributors may be used to endorse or promote products derived
 *   from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "emane/events/tdmascheduleevent.h"
#include "tdmascheduleevent.pb.h"

#include <cstdint>
#include <tuple>

class EMANE::Events::TDMAScheduleEvent::Implementation
{
public:
  Implementation(const Serialization & serialization):
    bHasStructure_{}
  {
    EMANEMessage::TDMAScheduleEvent msg{};

    if(!msg.ParseFromString(serialization))
      {
        throw SerializationException("unable to deserialize : TDMAScheduleEvent");
      }

    std::uint32_t u32FramesPerMultiFrame{};
    std::uint32_t u32SlotsPerFrame{};

    if(msg.has_structure())
      {
        const auto & structure = msg.structure();

        u32FramesPerMultiFrame = structure.framespermultiframe();
        u32SlotsPerFrame = structure.slotsperframe();

        structure_ = SlotStructure{structure.bandwidthhz(),
                                   u32FramesPerMultiFrame,
                                   u32SlotsPerFrame,
                                   Microseconds{structure.slotdurationmicroseconds()},
                                   Microseconds{structure.slotoverheadmicroseconds()}};

        bHasStructure_ = true;

        slotInfos_.reserve(u32FramesPerMultiFrame * u32SlotsPerFrame);
      }

    if(bHasStructure_)
      {
        // pre-built a complete schedule defaulting all slots to idle
        for(unsigned i =  0; i < u32FramesPerMultiFrame; ++i)
          {
            for(unsigned j =  0; j < u32SlotsPerFrame; ++j)
              {
                slotInfos_.push_back({SlotInfo::Type::IDLE,i,j});
              }
          }
      }

    for(const auto & frame :msg.frames())
      {
        std::uint32_t u32FrameIndex = frame.index();

        if(bHasStructure_ && u32FrameIndex >= u32FramesPerMultiFrame)
          {
            throw makeException<SerializationException>("TDMAScheduleEvent : Frame %lu index out of range",
                                                        u32FrameIndex);
          }

        std::set<std::uint32_t> presentSlots{};

        for(const auto & slot : frame.slots())
          {
            SlotInfo::Type type{SlotInfo::Type::IDLE};
            std::uint64_t u64FrequencyHz{};
            std::uint64_t u64DataRatebps{};
            std::uint8_t u8ServiceClass{};
            double dPowerdBm{};
            NEMId destination{};

            std::uint32_t u32SlotIndex = slot.index();

            if(bHasStructure_ && u32SlotIndex >= u32SlotsPerFrame)
              {
                 throw makeException<SerializationException>("TDMAScheduleEvent : Frame %lu Slot %lu slot index out of range",
                                                             u32FrameIndex,
                                                             u32SlotIndex);
              }

            presentSlots.insert(u32SlotIndex);

            switch(slot.type())
              {
              case EMANEMessage::TDMAScheduleEvent::Frame::Slot::SLOT_TX:
                {
                  const auto & tx = slot.tx();

                  type = SlotInfo::Type::TX;

                  if(tx.has_frequencyhz())
                    {
                      u64FrequencyHz = tx.frequencyhz();
                    }
                  else if(frame.has_frequencyhz())
                    {
                      u64FrequencyHz = frame.frequencyhz();
                    }
                  else if(msg.has_frequencyhz())
                    {
                      u64FrequencyHz = msg.frequencyhz();
                    }
                  else
                    {
                      throw makeException<SerializationException>("TDMAScheduleEvent : Frame %lu Slot %lu has undeterminable frequency",
                                                                  u32FrameIndex,
                                                                  u32SlotIndex);
                    }

                  if(tx.has_dataratebps())
                    {
                      u64DataRatebps = tx.dataratebps();
                    }
                  else if(frame.has_dataratebps())
                    {
                      u64DataRatebps = frame.dataratebps();
                    }
                  else if(msg.has_dataratebps())
                    {
                      u64DataRatebps = msg.dataratebps();
                    }
                  else
                    {
                      throw makeException<SerializationException>("TDMAScheduleEvent : Frame %lu Slot %lu has undeterminable datarate",
                                                                  u32FrameIndex,
                                                                  u32SlotIndex);
                    }


                  if(tx.has_serviceclass())
                    {
                      u8ServiceClass = tx.serviceclass();
                    }
                  else if(frame.has_serviceclass())
                    {
                      u8ServiceClass = frame.serviceclass();
                    }
                  else if(msg.has_serviceclass())
                    {
                      u8ServiceClass = msg.serviceclass();
                    }
                  else
                    {
                      throw makeException<SerializationException>("TDMAScheduleEvent : Frame %lu Slot %lu has undeterminable class",
                                                                  u32FrameIndex,
                                                                  u32SlotIndex);
                    }


                  if(tx.has_powerdbm())
                    {
                      dPowerdBm = tx.powerdbm();
                    }
                  else if(frame.has_powerdbm())
                    {
                      dPowerdBm = frame.powerdbm();
                    }
                  else if(msg.has_powerdbm())
                    {
                      dPowerdBm = msg.powerdbm();
                    }
                  else
                    {
                      throw makeException<SerializationException>("TDMAScheduleEvent : Frame %lu Slot %lu has undeterminable power",
                                                                  u32FrameIndex,
                                                                  u32SlotIndex);
                    }


                  if(tx.has_destination())
                    {
                      destination = tx.destination();
                    }
                }
                break;

              case EMANEMessage::TDMAScheduleEvent::Frame::Slot::SLOT_RX:
                {
                  const auto & rx = slot.rx();

                  type = SlotInfo::Type::RX;

                  if(rx.has_frequencyhz())
                    {
                      u64FrequencyHz = rx.frequencyhz();
                    }
                  else if(frame.has_frequencyhz())
                    {
                      u64FrequencyHz = frame.frequencyhz();
                    }
                  else if(msg.has_frequencyhz())
                    {
                      u64FrequencyHz = msg.frequencyhz();
                    }
                  else
                    {
                      throw makeException<SerializationException>("TDMAScheduleEvent : Frame %lu Slot %lu has undeterminable frequency",
                                                                  u32FrameIndex,
                                                                  u32SlotIndex);
                    }
                }
                break;

              case EMANEMessage::TDMAScheduleEvent::Frame::Slot::SLOT_IDLE:
                type = SlotInfo::Type::IDLE;
                break;
              }


            frequencies_.insert(u64FrequencyHz);

            if(bHasStructure_)
              {
                slotInfos_[u32FrameIndex * u32SlotsPerFrame + u32SlotIndex] =
                  {type,
                   u32FrameIndex,
                   u32SlotIndex,
                   u64FrequencyHz,
                   u64DataRatebps,
                   u8ServiceClass,
                   dPowerdBm,
                   destination};
              }
            else
              {
                slotInfos_.push_back({type,
                      u32FrameIndex,
                      u32SlotIndex,
                      u64FrequencyHz,
                      u64DataRatebps,
                      u8ServiceClass,
                      dPowerdBm,
                      destination});
              }
          }

        if(bHasStructure_)
          {
            // add RX slot info for any missing slot
            for(unsigned i = 0; i < u32SlotsPerFrame; ++i)
              {
                if(!presentSlots.count(i))
                  {
                    std::uint64_t u64FrequencyHz{};

                    if(frame.has_frequencyhz())
                      {
                        u64FrequencyHz = frame.frequencyhz();
                      }
                    else if(msg.has_frequencyhz())
                      {
                        u64FrequencyHz = msg.frequencyhz();
                      }
                    else
                      {
                        throw makeException<SerializationException>("TDMAScheduleEvent : Frame %lu Slot %lu has undeterminable frequency",
                                                                    u32FrameIndex,
                                                                    i);
                      }

                    frequencies_.insert(u64FrequencyHz);

                    slotInfos_[u32FrameIndex * u32SlotsPerFrame + i] = {SlotInfo::Type::RX,
                                                                        u32FrameIndex,
                                                                        i,
                                                                        u64FrequencyHz};
                  }
              }
          }
      }
  }

  const SlotInfos & getSlotInfos() const
  {
    return slotInfos_;
  }

  const Frequencies & getFrequencies() const
  {
    return frequencies_;
  }

  std::pair<const SlotStructure &,bool> getSlotStructure() const
  {
    return {structure_,bHasStructure_};
  }

private:
  SlotInfos slotInfos_;
  Frequencies frequencies_;
  SlotStructure structure_;
  bool bHasStructure_;
};


EMANE::Events::TDMAScheduleEvent::TDMAScheduleEvent(const Serialization & serialization):
  Event{IDENTIFIER},
  pImpl_{new Implementation{serialization}}{}

EMANE::Events::TDMAScheduleEvent::~TDMAScheduleEvent(){}

const EMANE::Events::SlotInfos & EMANE::Events::TDMAScheduleEvent::getSlotInfos() const
{
  return pImpl_->getSlotInfos();
}


const EMANE::Events::TDMAScheduleEvent::Frequencies &
EMANE::Events::TDMAScheduleEvent::getFrequencies() const
{
  return pImpl_->getFrequencies();
}

std::pair<const EMANE::Events::SlotStructure &,bool>
EMANE::Events::TDMAScheduleEvent::getSlotStructure() const
{
  return  pImpl_->getSlotStructure();
}
